### DISCUSSION TITLE
Proxyless GRPC Support

### DISCUSSION THREAD

**Discussion by altaiezior** [linkerd/linkerd2#9434] [Ideas]
### What problem are you trying to solve?

Adding support for proxyless grpc. With this support the application can bypass proxy and achieves load balancing capabilities.

### How should the problem be solved?

Grpc has support for xds api. Few other service mesh solutions provides this support and it becomes useful for high throughput latency critical environments.

### Any alternatives you've considered?

I haven’t given a spin up to linkerd yet, but I will be submitting performance comparison between linkerd proxy and grpc proxyless.

### How would users interact with this feature?

Not sure how linkerd plans to tackle this problem but there are few suggestions.

Proxyless grpc needs a bootstrap file to tell where control plane exists and how to interact with it. If we can provide the pod with proper bootstrap file that should be more or less sorted.

Issue will come in making other things streamlined like for example metrics and tracing since the traffic now doesn’t pass through linkerd. That is to be thought upon and at present even today there doesn’t exists a proper solution to resolve this.

### Would you like to work on this feature?

yes

**Comment by olix0r** [comment]
Linkerd won't interfere with load balancing decisions made by an application, but the Linkerd proxy does a lot more than load balancing.

It sounds to me like you're proposing implementing a control plane that provides the XDS API? If so, it sounds like you probably are better off building that outside of Linkerd.

If you're going to pursue this, it would be helpful to get clear about what, specifically, you want Linkerd to do for users in this configuration.

**Comment by altaiezior** [comment]
@olix0r thanks for taking time to reply back, this may sound like something where linkerd can't help, but let me state few points

1. A user can still use linkerd for ingress / egress traffic
2. A user would still require a control plane to manage the traffic directions, virtual service definitions etc. XDS API to grpc XDS enabled endpoints can use this data to redirect traffic efficiently without needing a proxy.
3. In a large cluster there can be a mix of different kinds of traffic including http 1.1, http 2, grpc and grpc with proxyless. In the first 3 scenarios it will be better to use a proxy while in the last scenario proxy can be bypassed.
4. For proxyless traffic since it is not going through the proxy it will be difficult to provide similar or all functionalities as it is also developing with time and enriching with more features. But few of the nice features can include connected observability, metrics. This will be difficult to achieve as I said because one needs to ensure that a metric is produced consistently across languages and the responsibility of producing these metrics will be part of the application.

There can more things, let's just first ponder upon the points above to see if this seems feasible with the roadmap of linkerd.

On a side note, we have run multiple benchmarks with grpc proxyless setup and plain istio envoy setup and proxyless has proved to be very efficient both in terms of resources and latency.

Istio envoy setup doesn't really work for our latency critical and high throughput traffic for sure. LinkerD I am yet to try, I was trying earlier but some issues came up while setting up.
